src/main/java/com/github/retro_game/retro_game/service/impl/ResetPasswordServiceImpl.java

```diff
@@ -0,0 +1,117 @@
+      @Value("${retro-game.password-reset-link-expires-after-days}") int passwordResetLinkExpiresAfterDays,
+      PasswordEncoder passwordEncoder,
+      UserRepository userRepository,
+      EmailService emailService,
+      UserPasswordResetTokenRepository userPasswordResetTokenRepository) {
+    this.passwordEncoder = passwordEncoder;
+    this.userRepository = userRepository;
+    this.emailService = emailService;
+    this.userPasswordResetTokenRepository = userPasswordResetTokenRepository;
+    this.passwordResetLinkExpiresAfterDays = passwordResetLinkExpiresAfterDays;
+  }
+
+  @Override
+  @Transactional
+  public void generateTokenAndSendEmail(String email) {
+    Optional<User> user = userRepository.findByEmailIgnoreCase(email);
+
+    user.ifPresent(u -> {
+      String plainToken = generateToken();
+      Optional<UserPasswordResetToken> existingToken = userPasswordResetTokenRepository.findByKey_User(u);
+
+      if (existingToken.isPresent()) {
+        userPasswordResetTokenRepository.save(
+            updatedToken(plainToken, existingToken.get())
+        );
+      } else {
+        userPasswordResetTokenRepository.save(
+            prepareNewToken(u, plainToken)
+        );
+      }
+
+      emailService.sendResetPasswordEmail(u, plainToken);
+    });
+  }
+
+  @Override
+  @Transactional
+  public void resetUserPassword(String plainToken, String newPassword) {
+    Optional<UserPasswordResetToken> token = userPasswordResetTokenRepository.findAll().stream()
+        .filter(t -> passwordEncoder.matches(plainToken, t.getEncryptedToken()))
+        .findFirst();
+
+    token.ifPresent(t -> {
+      if (!isTokenExpired(t)) {
+        updatePassword(newPassword, t);
+      } else {
+        userPasswordResetTokenRepository.delete(t);
+      }
+    });
+
+  }
+
+  private void updatePassword(String newPassword, UserPasswordResetToken t) {
+    User user = t.getKey().getUser();
+    user.setPassword(passwordEncoder.encode(newPassword));
+
+    userRepository.save(user);
+    userPasswordResetTokenRepository.delete(t);
+  }
+
+  private boolean isTokenExpired(UserPasswordResetToken token) {
+    return token.getExpireAt().before(new Date());
+  }
+
+  private UserPasswordResetToken updatedToken(String plainToken, UserPasswordResetToken existingToken) {
+    existingToken.setEncryptedToken(getEncryptedToken(plainToken));
+    return existingToken;
+  }
+
+  private UserPasswordResetToken prepareNewToken(User user, String plainToken) {
+    return new UserPasswordResetToken(
+        new UserPasswordResetTokenKey(user),
+        getEncryptedToken(plainToken),
+        calculateExpiredAt()
+    );
+  }
+
+  private String getEncryptedToken(String plainToken) {
+    return passwordEncoder.encode(plainToken);
+  }
+
+  private Date calculateExpiredAt() {
+    return DateUtils.addDays(new Date(), passwordResetLinkExpiresAfterDays);
+  }
+
+  private static String generateToken() {
+    return RandomStringUtils.randomAlphanumeric(30);
```

Comment:
`RandomStringUtils` is not cryptographically secure. We probably should use something safer for the tokens. Maybe, `UUID.randomUUID().toString()` might work.